source/darwin/securityframework_rsa.c

```diff
@@ -0,0 +1,543 @@
+    AWS_FATAL_ASSERT(error == NULL && "This function should not fail on a key we just generated");
+
+    struct aws_byte_cursor key_cur =
+        aws_byte_cursor_from_array(CFDataGetBytePtr(sec_key_export_data), CFDataGetLength(sec_key_export_data));
+
+    aws_byte_buf_init_copy_from_cursor(&key_pair->base.priv, allocator, key_cur);
+
+    sec_key_export_data = SecKeyCopyExternalRepresentation(key_pair->pub_key_ref, &error);
+    AWS_FATAL_ASSERT(error == NULL && "This function should not fail on a key we just generated");
+
+    key_cur = aws_byte_cursor_from_array(CFDataGetBytePtr(sec_key_export_data), CFDataGetLength(sec_key_export_data));
+
+    aws_byte_buf_init_copy_from_cursor(&key_pair->base.pub, allocator, key_cur);
+
+    key_pair->base.vtable = &s_rsa_key_pair_vtable;
+    key_pair->base.key_size_in_bits = key_size_in_bits;
+
+    CFRelease(sec_key_export_data);
+    CFRelease(cf_key_size);
+    CFRelease(key_attributes);
+
+    return &key_pair->base;
+
+on_error:
+    if (key_attributes) {
+        CFRelease(key_attributes);
+    }
+
+    if (sec_key_export_data) {
+        CFRelease(sec_key_export_data);
+    }
+
+    if (cf_key_size) {
+        CFRelease(cf_key_size);
+    }
+
+    s_rsa_destroy_key(&key_pair->base);
+    return NULL;
+}
+
+struct aws_rsa_key_pair *aws_rsa_key_pair_new_from_private_key_pkcs1_impl(
+    struct aws_allocator *allocator,
+    struct aws_byte_cursor key) {
+    struct sec_rsa_key_pair *key_pair_impl = aws_mem_calloc(allocator, 1, sizeof(struct sec_rsa_key_pair));
+
+    CFMutableDictionaryRef key_attributes = NULL;
+    CFDataRef private_key_data = NULL;
+
+    aws_ref_count_init(&key_pair_impl->base.ref_count, &key_pair_impl->base, s_rsa_destroy_key);
+    key_pair_impl->base.impl = key_pair_impl;
+    key_pair_impl->base.allocator = allocator;
+    key_pair_impl->cf_allocator = aws_wrapped_cf_allocator_new(allocator);
+    aws_byte_buf_init_copy_from_cursor(&key_pair_impl->base.priv, allocator, key);
+
+    private_key_data = CFDataCreate(key_pair_impl->cf_allocator, key.ptr, key.len);
+    AWS_FATAL_ASSERT(private_key_data);
+
+    key_attributes = CFDictionaryCreateMutable(key_pair_impl->cf_allocator, 0, NULL, NULL);
+    AWS_FATAL_ASSERT(key_attributes);
+   
+    CFDictionaryAddValue(key_attributes, kSecClass, kSecClassKey);
+    CFDictionaryAddValue(key_attributes, kSecAttrKeyType, kSecAttrKeyTypeRSA);
+    CFDictionaryAddValue(key_attributes, kSecAttrKeyClass, kSecAttrKeyClassPrivate);
+
+    CFErrorRef error = NULL;
+    key_pair_impl->priv_key_ref = SecKeyCreateWithData(private_key_data, key_attributes, &error);
+    if (s_reinterpret_sec_error_as_crt(error, "SecKeyCreateWithData"))  {
+        CFRelease(error);
+        goto on_error;
+    }
+
+    key_pair_impl->pub_key_ref = SecKeyCopyPublicKey(key_pair_impl->priv_key_ref);
+    if (key_pair_impl->pub_key_ref == NULL) {
+        aws_raise_error(AWS_ERROR_CAL_CRYPTO_OPERATION_FAILED);
+        CFRelease(error);
```

Comment:
don't free `error` here

and maybe raise MISSING_COMPONENT instead of generic FAILED? (I really don't know)

and log about it
```suggestion
    key_pair_impl->pub_key_ref = SecKeyCopyPublicKey(key_pair_impl->priv_key_ref);
    if (key_pair_impl->pub_key_ref == NULL) {
        AWS_LOGF_ERROR(AWS_LS_CAL_RSA, "Cannot create RSA key-pair, public key is missing.");
        aws_raise_error(AWS_ERROR_CAL_MISSING_REQUIRED_KEY_COMPONENT);
```

Reply:
switched to fatal assert. getting public key should never fail if loading private one suceeded